### ISSUE TITLE
checkpoint_required

### ISSUE THREAD

**Issue by oidevelop** [liamcottle/Instagram-SDK-PHP#50]
I can not authenticate and no instagram accounts with the sdk, I get the error:

Login Failed: [fail] checkpoint_required Go to this URL in your web browser to continue: https://i.instagram.com/challenge/[alphanumeric code]/

Possible solutions?

**Comment by ruslanbakanov** [comment]
Hello! :)
I have the same error since december 7

**Comment by shkumbinhasani** [comment]
i think this is because the device id get regenerated every login too if u verify the last login but i dont know how to add the same device too times i tried to save the session to verify and login with the session but it didnt work

**Comment by magicmarc** [comment]
Keeping the device ID constant (or the seed only based on username) doesn’t seem to help. Visiting the challenge URL or tapping “this was me” in the Instagram app causes the same issue on the next login attempt. 

I also kept the UUID and GUID constant and am getting the same issue. Is there anything else that can be kept constant so that the challenge allows a successful login?

**Comment by mewlabs** [comment]
try to use private proxy, it may help (but first time on new IP you might again receive checkpoint issue, but after confirmation it should work well)

**Comment by muthu-kc** [comment]
call checkpoint URL with cookies and send email verification code with the same URL. Finally, you will get session and cookies.

**Comment by Koldrish** [comment]
Thanks muthu-kc but I did not understand, you could give an example using code
PHP? You would be very helpful to many of us

**Comment by muthu-kc** [comment]
I wrote few classes with @liamcottle code. Modified library code [instagram-sdk-php.zip](https://github.com/liamcottle/Instagram-SDK-PHP/files/1777219/instagram-sdk-php.zip) and here's code how to use it.

Below code will send a verification code to email. If any other verification method needed, you can debug $verification_methods and choose on your own.

Depending on the value you choose for the challenge, it will send the code to your mobile or email.

```
            $response = $instagram->login($request->username, $request->password);

            if(!is_object($response) && isset($response['code']) && $response['code'] == 201){

              $url = $response['url'];

              $res = $instagram->ChallengeCode($response['url']);

              $pattern = '/window._sharedData = (.*);/';
              preg_match($pattern, $res, $matches);

              $verification_methods = json_decode($matches[1]);

              $default_method = $verification_methods->entry_data->Challenge[0]->extraData->content[3]->fields[0]->values[0];

              $response = $instagram->sendVerificationCode($response['url'], $default_method->value);

              $insta = $instagram->saveSession();
       }
```

Once you get verification code and submit code with below function. Note: You have initialized the save session from the previous method, which one will have csrf token, other cookies.

 ```
$instagram->initFromSavedSession($session);
      $response = $instagram->ConfirmVerificationCode($request->url, $request->code);

      if($response->status == "ok"){

        $savedSession = json_decode($instagram->saveSession());
        $userId  = $savedSession->cookies->ds_user_id;
        $savedSession->loggedInUser['pk'] = $userId;
      }
```

Finally, You have to set logged in user id manually, I can't figure out where the code exists for this. So this is the way to do it.

Let me know if it works. Make sure use same proxy for login process and usage.

**Comment by shkumbinhasani** [comment]
@muthu-kc can u too add the new post multiple photos feature i tried it but couldn't do it

**Comment by muthu-kc** [comment]
@shkumbinhasani   Do you know the endpoint and parameters we need to pass?

**Comment by shkumbinhasani** [comment]
@muthu-kc can u contact me my number is on my bio on whatsapp

**Comment by Vishal-sagar** [comment]
@muthu-kc  when i tried to saveSession() string in database and when using initFromSavedSession();
it says failed to deserialise session

**Comment by muthu-kc** [comment]
@Vishal-sagar Make sure you request is successful before saving the session.

**Comment by Vishal-sagar** [comment]
@muthu-kc  Yes actually i am able to save data in DB , but when i tried that text on json parser in cookies object it failed , i think it generates invalid json.

**Comment by muthu-kc** [comment]
use JSON encode before saving it to the database.

**Comment by Vishal-sagar** [comment]
@muthu-kc  saveSession() already return json_encode($session); data. but when i use initFromSavedSession() there is $session = json_decode($session, true);
it failed to decode json.

**Comment by muthu-kc** [comment]
@Vishal-sagar Did you checked the saved session data? is it valid one?

**Comment by Vishal-sagar** [comment]
@muthu-kc  Yes i tried to parse it on http://json.parser.online.fr/ , but it says invalid json , particularly on Cookies area.

**Comment by muthu-kc** [comment]
@Vishal-sagar Need your JSON

**Comment by Vishal-sagar** [comment]
@muthu-kc  Sure , give me sometime , i will post here

**Comment by Vishal-sagar** [comment]
@muthu-kc  Below is the json

{"loggedInUser":{"username":"vishal","phone_number":"","has_anonymous_profile_picture":false,"is_business":false,"profile_pic_url":"https://scontent-ort2-1.cdninstagram.com/vp/bc6a291dbdce559595fc975096409929/5B30907C/t51.2885-19/s150x150/28153685_156529775054712_47xxxx151424_n.jpg","full_name":"vishal","pk":"7xxxxxx52","is_verified":false,"is_private":false},"cookies":{"ds_user":"vishal","social_hash_bucket_id":"666","rur":"PRN","csrftoken":"yiwodwvF1eD5wBwxBtMYiL9StLi0zJDY","mid":"WqkOvAABAAGiSrfjvi_eJpwItqjJ","ds_user_id":"7xxxxxxx52","urlgen":""{\"time\": 1521028797\054 \"67.209.122.237\": 55293}:1ew54X:G59zFk_GOUhY2BrQpTDbzRBUe18"","sessionid":"IGSCeb980ef7c54449155778f45a5c6061d43be2d818ec65438d109027a968475bfc%3AdteGNZXhZeZrlXmZ5zWfG7YUwPyXGHPs%3A%7B%22_auth_user_id%22%3A7186133352%2C%22_auth_user_backend%22%3A%22accounts.backends.CaseInsensitiveModelBackend%22%2C%22_auth_user_hash%22%3A%22%22%2C%22_platform%22%3A1%2C%22_token_ver%22%3A2%2C%22_token%22%3A%227186133352%3AEnt7ysXLkIPdu1zePdWeG9vbmOiV2RW5%3A6c52a0adc7296ac715db1aff66ff8dc19be81216514da913f44d892bea1d7653%22%2C%22last_refreshed%22%3A1521028797.3432145119%7D"},"csrfToken":"yiwodwvF1eD5wBwxBtMYiL9StLi0zJDY","rankToken":"97cd6931-8efb-416d-8799-525a664e97a0","phoneId":"8e5ca1dc-8dcf-4293-a01b-afe4f056f3a7","deviceId":"android-df1ff0c22e3b9aca","guid":"add3d995-7bf0-43a1-bb31-8b1c3b7430da","uuid":"add3d995-7bf0-43a1-bb31-8b1c3b7430da","googleAdId":"8c7b3117-dbe6-47e8-b72f-5d431405e1c9"}

**Comment by muthu-kc** [comment]
@Vishal-sagar Sorry I can't figure out how this problem occurs. Are you sure you are using the correct code? I only posted the code here for checkpoint verification. But you have JSON for the authorized user.

**Comment by Vishal-sagar** [comment]
@muthu-kc  ,Here is the flow of my code-> First i do Login with username/Password then saveSession() and store to db and next time initFromSavedSession().

**Comment by Vishal-sagar** [comment]
@muthu-kc  i am doing for Normal login not for CheckPoint. i am using sdk to do autoPost from server so i want to use session so no need of Frequently Login and Little bit safe From blockage from instagram.

**Comment by muthu-kc** [comment]
@Vishal-sagar  Try to use saved session immediately before saving to the database. If it works, make sure you save json in the correct format to the database. Because in your JSON there's some escape thing happened. Please message on Twitter, if you still have issue. https://twitter.com/muthu_phpian

**Comment by Vishal-sagar** [comment]
@muthu-kc  Thank you for the support.
it works , issue was when saving on DB it messed up json. so saved in text file and saved that file to DB. and it works.
One more question , Do you have any solution to Save Server Blockage from Instagram ?

**Comment by Koldrish** [comment]
does anyone update a fork with all these changes?

**Comment by henzan02** [comment]
Hi @muthu-kc I have following your code and I do this : 

```
if($response->status == "ok"){

    $savedSession = json_decode($instagram->saveSession());
    $userId  = $savedSession->cookies->ds_user_id;
    $savedSession->loggedInUser['pk'] = $userId;
    $new = json_encode($savedSession);    
    $newSession = serialize(urlencode($new));    
    print_r($newSession);
}
```

Then I've got : 

`{"loggedInUser":{"pk":"7071427002"},"cookies":{"rur":"ASH","csrftoken":"5Byssl6htcwdADS6xkdnWInyEKzTynRj","mid":"Wr2h4wABAAFsVJTh9Khakjejdp4X","urlgen":"\"{\\\"time\\\": 1522377187\\054 \\\"119.81.196.46\\\": 36351}:1f1jql:0FqjKVPn94bu7N1fCp7ckWrgsbU\"","ds_user_id":"7071427002","sessionid":"IGSCb67c75758c969e0eddef3ae68d57b37b7adadcd11a4211aa07c67492e1e7f442%3AWSDevgQEy7tjplQWi8Bj9D8RBkvMTr7C%3A%7B%22_auth_user_id%22%3A7071427002%2C%22_auth_user_backend%22%3A%22accounts.backends.CaseInsensitiveModelBackend%22%2C%22_auth_user_hash%22%3A%22%22%2C%22_platform%22%3A4%2C%22_token_ver%22%3A2%2C%22_token%22%3A%227071427002%3AHTScq1OrrDIB5Z1GQzl0fy2raZXloHqO%3A134483b20caf2b7c7cd2c15e726638ad59a5cfeb18dbc366f1160361a973e666%22%2C%22last_refreshed%22%3A1522377247.2425196171%7D"},"csrfToken":"5Byssl6htcwdADS6xkdnWInyEKzTynRj","rankToken":"a9e3f2d2-04f5-4192-a02d-2d02b9771baa","phoneId":"42b058ab-57ef-482f-ac3a-fe4b7232666a","deviceId":"android-0fbbecfc9dcb502d","guid":"17002f79-c694-45a5-ad8d-85de5b0d51e4","uuid":"17002f79-c694-45a5-ad8d-85de5b0d51e4","googleAdId":"0f873f6b-ed5c-49d1-b493-36284cc96f1a"}`

So what next?

Please help. 

Thanks

**Comment by muthu-kc** [comment]
Are you new to development? Save the `$new` JSON value to the database and use it to initiate session again where ever you want `$instagram->initFromSavedSession($new);` Now you can do your actions.

**Comment by mikhailfarberov** [comment]
@muthu-kc , I've found this thread when faced the same issue. Your solution works but in the result after sending a code i get this response and no user id in cookies:
RESPONSE:
object(stdClass)#28 (3) { ["location"]=> string(24) "https://i.instagram.com/" ["type"]=> string(21) "CHALLENGE_REDIRECTION" ["status"]=> string(2) "ok" }

COOKIES:
 object(stdClass)#17 (9) { ["loggedInUser"]=> array(0) { } ["cookies"]=> object(stdClass)#29 (4) { ["rur"]=> string(3) "ATN" ["csrftoken"]=> string(32) "hQ0MQfzNo4zm1YIU57DImI0rx133xHO7" ["mid"]=> string(28) "WuXpWwABAAEzxdJQuLmOZ9eI_UEQ" ["urlgen"]=> string(89) ""{\"time\": 1525016922\054 \"77.244.221.148\": 49505}:1fCoZ8:PhDHH1cj3_eAvjHnJNHabcH2yhQ"" } ["csrfToken"]=> string(32) "hQ0MQfzNo4zm1YIU57DImI0rx133xHO7" ["rankToken"]=> string(36) "294ee666-c769-499f-9789-46198e02bbca" ["phoneId"]=> string(36) "3149c64c-4bb3-4e81-8582-e5c61e0fb105" ["deviceId"]=> string(24) "android-53e6b4fbd3202f81" ["guid"]=> string(36) "d8132975-4113-461d-86fb-e29ea8870737" ["uuid"]=> string(36) "d8132975-4113-461d-86fb-e29ea8870737" ["googleAdId"]=> string(36) "3c5e8c63-1b68-4165-8c0d-12172e30f47b" }

Further attempts to call instagram methods fail, seems like my account is still not logged in.

Have you got the same problems?

**Comment by muthu-kc** [comment]
@mikhailfarberovIt happens when you use a public proxy or perform too many actions in short time. But at first time usually, this error will not happen.
Try using a private proxy that's located in your current location.

**Comment by mikhailfarberov** [comment]
@muthu-kc , thank you! I ve waited for 24 hours and now my ip is not blocked. Everything works fine.

**Comment by muthu-kc** [comment]
@mikhailfarberov  Thank you for letting me know. I will update this information.

**Comment by muthu-kc** [comment]
I created example for checkpoint issue.

https://github.com/muthu-kc/php-instagram-private-api-with-checkpoint-feature
